builtin_array.go

```diff
@@ -1224,6 +1224,169 @@ func (r *Runtime) arrayproto_flatMap(call FunctionCall) Value {
+		} else {
+			fromValue = o.self.getIdx(pk, nil)
+		}
+		createDataPropertyOrThrow(a, pk, fromValue)
+	}
+	return a
+}
+
+func (r *Runtime) arrayproto_toReversed(call FunctionCall) Value {
+	o := call.This.ToObject(r)
+	length := toLength(o.self.getStr("length", nil))
+	a := r.newArrayLength(length)
+
+	if src := r.checkStdArrayObj(o); src != nil {
+		for k := int64(0); k < length; k++ {
+			pk := valueInt(k)
+			from := valueInt(length - k - 1)
+			fromValue := src.values[from]
+			createDataPropertyOrThrow(a, pk, fromValue)
+		}
+	} else {
+		for k := int64(0); k < length; k++ {
+			pk := valueInt(k)
+			from := valueInt(length - k - 1)
+			fromValue := o.self.getIdx(from, nil)
+			createDataPropertyOrThrow(a, pk, fromValue)
+		}
+	}
+
+	return a
+}
+
+func (r *Runtime) arrayproto_toSorted(call FunctionCall) Value {
+	var compareFn func(FunctionCall) Value
+	arg := call.Argument(0)
+	if arg != _undefined {
+		if arg, ok := arg.(*Object); ok {
+			compareFn, _ = arg.self.assertCallable()
+		}
+		if compareFn == nil {
+			panic(r.NewTypeError("The comparison function must be either a function or undefined"))
+		}
+	}
+
+	o := call.This.ToObject(r)
+	length := toLength(o.self.getStr("length", nil))
+	if length >= math.MaxUint32 {
+		panic(r.newError(r.getRangeError(), "Invalid array length"))
+	}
+	a := make([]Value, 0, length)
+
+	for i := int64(0); i < length; i++ {
+		idx := valueInt(i)
+		a = append(a, nilSafe(o.self.getIdx(idx, nil)))
+	}
+	ar := r.newArrayValues(a)
+	ctx := arraySortCtx{
+		obj:     ar.self,
+		compare: compareFn,
+	}
+
+	sort.Stable(&ctx)
+	return ar
+}
+
+func (r *Runtime) arrayproto_toSpliced(call FunctionCall) Value {
+	o := call.This.ToObject(r)
+	length := toLength(o.self.getStr("length", nil))
+	actualStart := relToIdx(call.Argument(0).ToInteger(), length)
+	var actualSkipCount int64
+	switch len(call.Arguments) {
+	case 0:
+	case 1:
+		actualSkipCount = length - actualStart
+	default:
+		actualSkipCount = min(max(call.Argument(1).ToInteger(), 0), length-actualStart)
+	}
+	itemCount := max(int64(len(call.Arguments)-2), 0)
+	newLength := length - actualSkipCount + itemCount
+	if newLength >= maxInt {
+		panic(r.NewTypeError("Invalid array length"))
+	}
+
+	if src := r.checkStdArrayObj(o); src != nil {
+		var values []Value
+		if itemCount < actualSkipCount {
+			values = src.values
+			copy(values[actualStart+itemCount:], values[actualStart+actualSkipCount:])
+			tail := values[newLength:]
+			for k := range tail {
+				tail[k] = nil
+			}
+			values = values[:newLength]
+		} else if itemCount > actualSkipCount {
+			if int64(cap(src.values)) >= newLength {
+				values = src.values[:newLength]
+				copy(values[actualStart+itemCount:], values[actualStart+actualSkipCount:length])
+			} else {
+				values = make([]Value, newLength)
+				copy(values, src.values[:actualStart])
+				copy(values[actualStart+itemCount:], src.values[actualStart+actualSkipCount:])
+			}
+		} else {
+			values = src.values
+		}
```

Comment:
Given that the whole point of toSpliced is to not touch the original array. The code here seems very risky and IMO buggy, just not in a way that the tc39 test catch it.

Try to add a test such as 
```golang

func TestArrayToSpliced(t *testing.T) {
	const SCRIPT = `
		const a = [1, 2, 3];
		a.push(4)
		assert(compareArray(a, [1, 2, 3, 4]));
		const b = a.toSpliced(2)
		$ERROR(a)
		assert(compareArray(a, [2, 3, 4]));
		// assert(compareArray(b, [2, 3, 4, 3]));
	`
	testScriptWithTestLib(SCRIPT, _undefined, t)
}
```

And you will notice that `a` is now `1,2,,,` even though it is supposed to not be changed

Reply:
Sorry, I forgot to check this part.